### ISSUE TITLE
Querying using IS with type that dont declare TypeName does not work

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2635]
```C#
[Database] public class Foo { [Type] public FooType Type; }
[Database] public class FooType {}
```

The above model works fail the following query:

```C#
Db.SQL("SELECT f FROM Foo f WHERE f IS ?", Db.TypeOf<Foo>());
```

The error states:

> System.ArgumentException: ScErrBadArguments (SCERR1001): One or more arguments was invalid. Object without type name defined cannot be used as a type

It can be worked around by changing the model to

```C#
[Database] public class Foo { [Type] public FooType Type; }
[Database] public class FooType {
  [TypeName] public string Name;
}
```

and hence is no showstopper. But it should be resolved eventually - the implicit name column can/should be used by the query processor.

**Comment by k-rus** [comment]
What is the stack trace of the error?
This is not a bug, it was implemented according to the specification. Originally it was decided that the type name is required, otherwise an object cannot be used as type.
I don't know how if the current temporal implementation relies on this.

Since the implementation is temporal I suggest to not put effort to change the implementation.

**Comment by per-samuelsson** [comment]
> This is not a bug, it was implemented according to the specification.

Yes, I know it's according to the original specification. But later we introduced the notion of implicit columns, and I think that the final implementation should eventually use that as a fallback if there are no custom class / reference defined.

And the same goes for #2636.

No prio for none of them, though. I basically just "record" anything I find that I think we eventually should improve on. We can label them enhancements instead.
